### PR TITLE
TW-2968(fix): wrong member count in group chat

### LINKED ISSUE #2968: Wrong member count in group chat
2.11.1, [[linagora.com](http://linagora.com/)](http://linagora.com), IOS 

The counter of members says there’s only 1 member in a chat, though there are a lllot of members there

Image

Update 26/03/2026

Diana got this issue too in the latest version. I think it happens in older chats in which we have this bug with all members displayed as “read-only” (maybe in member count we don’t include read-only users? But I have such chats on my side and the member count is correct…)

Image

Image

### PR DESCRIPTION
**Pull request by jbrnzl**
## Ticket

Fixes #2968 

## Root cause & Context

We previously added a sentry log #2970 member to follow #2968 and I was wondering if we still had the issue or not and we have.
I made some investigation and indeed the member count comes from the room summary, and it looks like servers only send `m.joined_member_count` `m.invited_member_count` in /sync when they change. If one update is missed (gapped sync, app killed mid-sync), the persisted counts stay wrong forever: nothing ever reconciles them with the member list.
I could experiment this case with few rooms/accounts:
photo_2026-06-12_17-20-16 photo_2026-06-12_17-20-24

And we can verify it debugging (the room in the last screenshot with 4 members):
As you can see matrix-dart-sdk is giving us 0 and 1 but the total of the members is 4.

Capture d’écran 2026-06-12 à 15 03 20

## Changes

New `requestParticipantsWithSummaryReconciliation` extension on `Room`: after fetching the participants, if the real counts differ from the summary, the corrected values are re-injected through the sync pipeline (memory + database + UI). Called when opening a chat. Each healing is logged with the existing `Wrong-member-count` Sentry tag so we can follow the fix in production.

##  Concerns

I know this is not a perfect fix as the bug is an edge case where some informations were not stored as I said in the "Root cause" but this is a try fixing the UI client side. We should not rely on self-syncing but this is the only path I found to fix it. Maybe a more complete fix should be handled server side or within the matrix-dart-sdk but I am not sure how exatcly for now. Please feel free to help me make sure we can safely introduce this fix by testing or discussing it.

## Test recommendations

Unit tests cover the reconciliation and its guards. Manually: open a group whose displayed count is wrong, then the details page, the count should match the member list, with a `Wrong-member-count: healing summary` line in the logs.

## Release Notes

* **Bug Fixes**
  * Improved chat room member count accuracy by automatically reconciling stored summary counts against the actual participant membership when requesting members (with reconciliation limited to join/invite scenarios and avoiding incorrect updates in edge cases).

* **Tests**
  * Added/expanded automated tests to verify reconciliation updates are applied only when needed, persisted exactly once when stale, and skipped when counts are already correct or reconciliation shouldn’t run.
  * Updated test client setup to allow optional database injection and streamlined transaction behavior.

### CODE REVIEW

File: lib/utils/matrix_sdk_extensions/room_member_count_reconciliation_extension.dart [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      '$id from (joined=${summary.mJoinedMemberCount}, '
+      'invited=${summary.mInvitedMemberCount}) to (joined=$joinedCount, '
+      'invited=$invitedCount)',
+    );
+    // Feed the corrected counts through the regular sync pipeline, as if the
+    // server had sent them: same merge (heroes preserved), same persistence
+    // and UI notifications as a real summary update.
+    await client.handleSync(
+      SyncUpdate(
+        nextBatch: client.prevBatch ?? '',
```
Comment (nqhhdev) [inline lib/utils/matrix_sdk_extensions/room_member_count_reconciliation_extension.dart]: If `client.prevBatch` is null, the synthetic SyncUpdate gets nextBatch: ''. The SDK may persist '' as the new sync token, causing the next real /sync to restart from the beginning (full re-sync). Unlikely in practice but a silent data hazard. WDYT?

**Comment by 9clg6** [comment]
I think your fix is good. The server only includes m.joined_member_count / m.invited_member_count in /sync when they change, that's spec-compliant, not a server bug. The problem is purely client-side: if we miss that single update (gapped sync, app killed mid-sync), the persisted count stays stale forever, because the server won't resend the summary until the next change. Recomputing/reconciling from the local member list is the right place to fix it.
One thing to double-check: we should only override with the recomputed count when the member list is fully loaded locally, otherwise (lazy-loaded / partial membership) the recount could be lower than reality.

**Review by 9clg6 — APPROVED** [review]

File: lib/utils/matrix_sdk_extensions/room_member_count_reconciliation_extension.dart (hunk 1 above)
Comment (jbrnzl) [reply lib/utils/matrix_sdk_extensions/room_member_count_reconciliation_extension.dart]: Thank you @nqhhdev ! I checked and from what I can see handleSync does not read nextBatch. I think it is only used for real /sync. Btw I changed it to `nextBatch: ''` as in other places in the project. Don't hesitate if you have any other concerns

**Comment by jbrnzl** [comment]
> I think your fix is good. The server only includes m.joined_member_count / m.invited_member_count in /sync when they change, that's spec-compliant, not a server bug. The problem is purely client-side: if we miss that single update (gapped sync, app killed mid-sync), the persisted count stays stale forever, because the server won't resend the summary until the next change. Recomputing/reconciling from the local member list is the right place to fix it. One thing to double-check: we should only override with the recomputed count when the member list is fully loaded locally, otherwise (lazy-loaded / partial membership) the recount could be lower than reality.

Thank your for the review ! Yes, I can confirm it is well handled in requestParticipants (with getMembersByRoom) 🙏🏻

**Comment by krille-chan** [comment]
@jbrnzl this looks actually like you are working around a bug in synapse (or a severe bug in the sdk?) 🤔 while it is the responsibility of the sync loop to always give you an updated count. Do you know how an "missed delta" can happen there and how to reproduce it? Would like to fix the root cause

**Comment by jbrnzl** [comment]
> @jbrnzl this looks actually like you are working around a bug in synapse (or a severe bug in the sdk?) 🤔 while it is the responsibility of the sync loop to always give you an updated count. Do you know how an "missed delta" can happen there and how to reproduce it? Would like to fix the root cause

@krille-chan I think the easier to reproduce this issue is to create a new room from user `@a:hs1` and invite `@b:hs2`.
On `@b:hs2` screen you will see a wrong member count.
Here is the exemple:

https://github.com/user-attachments/assets/2dff9964-76e7-4d84-888d-c7c89d9ee999

Reproductible on FluffyChat and most of matrix clients I think as they behave the same:

photo_2026-06-17_14-22-21

Member count seems to never updates until there is a new `member event` in the room

**Comment by krille-chan** [comment]
@jbrnzl oh wow very good catch. I was able to reproduce it but also could confirm that this is not a bug in the sdk but just coming down from synapse. I have opened a bug report at synapse here: https://github.com/element-hq/synapse/issues/19865

**Comment by jbrnzl** [comment]
I think we can merge this fix event if it seems like a synapse issue. We can still keep track of "wrongMemberCount" on sentry, and see later if we can remove this fix, or if we have an other issue with the member count when the synapse issue is resolved.
What do you think @9clg6 @nqhhdev ?

**Comment by jbrnzl** [comment]
> @jbrnzl oh wow very good catch. I was able to reproduce it but also could confirm that this is not a bug in the sdk but just coming down from synapse. I have opened a bug report at synapse here: [element-hq/synapse#19865](https://github.com/element-hq/synapse/issues/19865)

Thank you very much !

**Comment by 9clg6** [comment]
@jbrnzl  Yes we can merge this workaround, tagged with a TODO waiting to a solution from Synapse.

**Review by 9clg6 — APPROVED** [review]

**Review by nqhhdev — APPROVED** [review]

**Review by 9clg6 — APPROVED** [review]
